src/rluni/robot/robot.py

```diff
@@ -0,0 +1,428 @@
+        except asyncio.CancelledError:
+            logger.info("Robot Loop Shutdown Signal.")
+            # Set the shutdown event to ensure control_loop exits
+            shutdown_event.set()
+            raise
+
+    async def control_loop(self, shutdown_event):
+        """
+        The main control loop for the robot. Executes sensor reading, state estimation, control decision making,
+        and actuator commands at a fixed rate defined by LOOP_TIME.
+        """
+        loop_period = self.LOOP_TIME
+        torque_request = 0
+
+        while not shutdown_event.is_set():
+            # Start Loop Timer and increment loop iteration
+            loop_start_time = time.time()
+            self.itr += 1
+
+            # Debugging Data Example (add more data as needed)
+            tele_debug_data = td.DebugData()
+            tele_debug_data.add_data(example_debug_data=42)
+
+            # Sensor reading and fusion
+
+            imudata = td.IMUData(*self.imu.read_sensor_data(convert=True))
+
+            quaternion, internal_states, flags = self.sensor_fusion.update(
+                imudata.get_gyro(),
+                imudata.get_accel(),
+                mag_data=imudata.get_mag(),
+                delta_time=loop_period,
+            )
+            rigid_body_state = td.EulerAngles(*self.sensor_fusion.euler_angles, *self.sensor_fusion.euler_rates)
+            
+            # Update robot state and parameters
+            # if self.xmotor is not None:
+            #     await self.xmotor.update_state()
+
+            for motor in self.motors:
+                if motor is not None: 
+                    await motor.update_state()
+
+            # Control logic
+            # control_input = ControlInput(
+            #     pendulum_angle=rigid_body_state.y * DEG_TO_RAD,  # [radians] positive CCW
+            #     pendulum_vel=imudata.gyro_z * DEG_TO_RAD,  # [radians / s] positive CCW
+            #     wheel_vel=(
+            #         0
+            #         if self.xmotor is None
+            #         else -self.xmotor.state["VELOCITY"] * REV_TO_RAD
+            #     ),  # [radians / s] positive CCW
+            #     roll_torque=torque_request,  # [N * m] positive CCW
+            # )
+
+            control_input = ControlInput(
+                euler_angles_x_rads=rigid_body_state.x * DEG_TO_RAD,
+                euler_angles_y_rads=rigid_body_state.y * DEG_TO_RAD,
+                euler_angles_z_rads=rigid_body_state.z * DEG_TO_RAD,
+                euler_rates_x_rads_s=rigid_body_state.x_dot,
+                euler_rates_y_rads_s=rigid_body_state.y_dot,
+                euler_rates_z_rads_s=rigid_body_state.z_dot,
+                motor_speeds_pitch_rads_s=self.motors.pitch.state["VELOCITY"]*REV_TO_RAD,
+                motor_speeds_roll_rads_s=self.motors.roll.state["VELOCITY"]*REV_TO_RAD,
+                motor_speeds_yaw_rads_s=self.motors.yaw.state["VELOCITY"]*REV_TO_RAD,
+            )
```

Comment:
This might be a nice teledata object to have. Not sure, can look at this again some other time